## Steps from the Exercise

In [2]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [3]:
# Importing orders data - using os library
path = r'C:\Users\35196\Documents\Alexandra\Data Analytics Program\02 Immersion\4. Python Fundamentals for Data Analysts\Instacart Basket Analysis'

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [4]:
# Create a dataframe
df_test = pd.DataFrame()

In [5]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [6]:
df_test

mix
0     a
1     b
2     1
3  True

In [8]:
# Check mixed-types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
# converte data type from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

# Converte data type from string to numeric 
# simply update astyper('str') to astype('int64')

In [12]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   mix     4 non-null      object
dtypes: object(1)
memory usage: 164.0+ bytes


In [13]:
# Finding Missing Values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
# After identify where is missing values and how much, let's create a subset of the dataframe containing only the values in question
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [7]:
df_prods.shape

(49693, 5)

In [9]:
# Create a clean dataframe, delete missing records
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]
df_prods_clean.shape

(49677, 5)

In [10]:
# Another way to drop all missing values is using dropna function - intead creating anew dataframe, this function overwrite df_prods (by inplace = True)
# Overwriting can be risky
# df_prods.dropna(subset = [‘product_name’], inplace = True)

In [11]:
# Finding Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [12]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [14]:
df_prods_clean.shape

(49677, 5)

In [15]:
# Deleting duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [16]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task

In [17]:
# Step 2 - run the df.describe()
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Based on describe() function I don't see any discripcancy on data.
I can spot count of days_since_prior_order is lower than the other columns, but this make sense, for those user that order for the first time
All the columns don't show anything to investigated further.

In [19]:
# Step 3 - check for mixed_type
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

No mixed-type columns were found

In [20]:
# Step 5 - Check for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Only the column days_since_prior_order has missing values, with 206209 records.
As I commented in step 1, these missing values are related with the first order for each user. Before the first order is impossible to calculate how many days happened the last/previous order because it never happened.



In [22]:
# Step 6 - Address the missing values using tan appropriate method
# In my opinion, the order dataframe should have the missing value, lets just create a subset countaining them 
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull() == True]
df_nan_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [23]:
# Step 7 - Run a check for duplicate values
df_dups_ords = df_ords[df_ords.duplicated()]

In [24]:
df_dups_ords

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates were found.

In [25]:
# Step 9 - Export cleaned df_prods and df_ords data as a '.csv'

In [27]:
# df_prods
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [26]:
#df_ords
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))